phpunit/html/wp-html-walker-test.php

```diff
@@ -0,0 +1,1012 @@
+		$w = new WP_HTML_Walker( self::HTML_WITH_CLASSES );
+		$w->next_tag();
+		$w->add_class( 'with-border' );
+		$this->assertSame(
+			'<div class="main with-border" id="first"><span class="not-main bold with-border" id="second">Text</span></div>',
+			(string) $w
+		);
+	}
+
+	/**
+	 * @ticket 56299
+	 */
+	public function test_preserves_class_name_order_when_a_duplicate_class_name_is_added() {
+		$w = new WP_HTML_Walker( self::HTML_WITH_CLASSES );
+		$w->next_tag();
+		$w->add_class( 'main' );
+		$this->assertSame(
+			'<div class="main with-border" id="first"><span class="not-main bold with-border" id="second">Text</span></div>',
+			(string) $w
+		);
+	}
+
+	/**
+	 * @ticket 56299
+	 */
+	public function test_add_class_when_there_is_a_class_attribute_with_excessive_whitespaces() {
+		$w = new WP_HTML_Walker(
+			'<div class="   main   with-border   " id="first"><span class="not-main bold with-border" id="second">Text</span></div>'
+		);
+		$w->next_tag();
+		$w->add_class( 'foo-class' );
+		$this->assertSame(
+			'<div class="   main   with-border foo-class" id="first"><span class="not-main bold with-border" id="second">Text</span></div>',
+			(string) $w
+		);
+	}
+
+	/**
+	 * @ticket 56299
+	 */
+	public function test_remove_class_when_there_is_a_class_attribute_with_excessive_whitespaces() {
+		$w = new WP_HTML_Walker(
+			'<div class="   main   with-border   " id="first"><span class="not-main bold with-border" id="second">Text</span></div>'
+		);
+		$w->next_tag();
+		$w->remove_class( 'with-border' );
+		$this->assertSame(
+			'<div class="   main" id="first"><span class="not-main bold with-border" id="second">Text</span></div>',
+			(string) $w
+		);
+	}
+
+	/**
+	 * @ticket 56299
+	 */
+	public function test_remove_all_classes_when_there_is_a_class_attribute_with_excessive_whitespaces() {
+		$w = new WP_HTML_Walker(
+			'<div class="   main   with-border   " id="first"><span class="not-main bold with-border" id="second">Text</span></div>'
+		);
+		$w->next_tag();
+		$w->remove_class( 'main' );
+		$w->remove_class( 'with-border' );
+		$this->assertSame(
+			'<div  id="first"><span class="not-main bold with-border" id="second">Text</span></div>',
+			(string) $w
+		);
+	}
+
+	/**
+	 * @ticket 56299
+	 */
+	public function test_set_attribute_takes_priority_over_add_class() {
+		$w = new WP_HTML_Walker( self::HTML_WITH_CLASSES );
+		$w->next_tag();
+		$w->add_class( 'add_class' );
+		$w->set_attribute( 'class', 'set_attribute' );
+		$this->assertSame(
+			'<div class="set_attribute" id="first"><span class="not-main bold with-border" id="second">Text</span></div>',
+			(string) $w
+		);
+
+		$w = new WP_HTML_Walker( self::HTML_WITH_CLASSES );
+		$w->next_tag();
+		$w->set_attribute( 'class', 'set_attribute' );
+		$w->add_class( 'add_class' );
+		$this->assertSame(
+			'<div class="set_attribute" id="first"><span class="not-main bold with-border" id="second">Text</span></div>',
+			(string) $w
+		);
+	}
+
+	/**
+	 * @ticket 56299
```

Comment:
A description for this advanced use case would be helpful to let others know what's being tested without having to analyse the whole test method.

Reply:
Explanation added in e0910be12f